public/app/features/alerting/unified/components/rule-editor/labels/LabelsField.tsx

```diff
@@ -0,0 +1,477 @@
+  dataSourceName: string;
+}
+export function LabelsWithSuggestions({ dataSourceName }: LabelsWithSuggestionsProps) {
+  const styles = useStyles2(getStyles);
+  const {
+    control,
+    watch,
+    formState: { errors },
+  } = useFormContext<LabelsSubformValues>();
+
+  const labelsInSubform = watch('labelsInSubform');
+  const { fields, remove, append } = useFieldArray({ control, name: 'labelsInSubform' });
+  const appendLabel = useCallback(() => {
+    append({ key: '', value: '' });
+  }, [append]);
+
+  // check if the labels plugin is installed
+  const { installed: labelsPluginInstalled = false, loading: loadingLabelsPlugin } = usePluginBridge(
+    SupportedPlugin.Labels
+  );
+  const [selectedKey, setSelectedKey] = useState('');
+
+  const { loading, labelsByKeyOps, keysFromExistingAlerts, labelsByKeyFromExisingAlerts, groupedOptions } =
+    useCombinedLabels(dataSourceName, labelsPluginInstalled, loadingLabelsPlugin, labelsInSubform);
+
+  const selectedKeyIsFromAlerts =
+    labelsByKeyFromExisingAlerts[selectedKey] !== undefined && labelsByKeyFromExisingAlerts[selectedKey]?.size > 0;
+  const selectedKeyIsFromOps = labelsByKeyOps[selectedKey] !== undefined && labelsByKeyOps[selectedKey]?.size > 0;
+  const selectedKeyDoesNotExist = !selectedKeyIsFromAlerts && !selectedKeyIsFromOps;
+
+  const valuesAlreadyFetched = !selectedKeyIsFromAlerts && labelsByKeyOps[selectedKey]?.size > 0;
+
+  // Only fetch the values for the selected key if it is from ops and the values are not already fetched (the selected key is not in the labelsByKeyOps object)
+  const {
+    currentData: valuesData,
+    isLoading: isLoadingValues = false,
+    error,
+  } = labelsApi.endpoints.getLabelValues.useQuery(
+    { key: selectedKey },
+    {
+      skip:
+        !labelsPluginInstalled ||
+        !selectedKey ||
+        selectedKeyIsFromAlerts ||
+        valuesAlreadyFetched ||
+        selectedKeyDoesNotExist,
+    }
+  );
+
+  // these are the values for the selected key in case it is from ops
+  const valuesFromSelectedGopsKey = useMemo(() => {
+    // if it is from alerts, we need to fetch the values from the existing alerts
+    if (selectedKeyIsFromAlerts) {
+      return [];
+    }
+    // in case of a label from ops, we need to fetch the values from the plugin
+    // fetch values from ops only if there is no value for the key
+    const valuesForSelectedKey = labelsByKeyOps[selectedKey];
+    const valuesAlreadyFetched = valuesForSelectedKey?.size > 0;
+    if (valuesAlreadyFetched) {
+      return mapLabelsToOptions(valuesForSelectedKey);
+    }
+    if (!isLoadingValues && valuesData?.values?.length && !error) {
+      const values = valuesData?.values.map((value) => value.name);
+      labelsByKeyOps[selectedKey] = new Set(values);
+      return mapLabelsToOptions(values);
+    }
+    return [];
+  }, [selectedKeyIsFromAlerts, labelsByKeyOps, selectedKey, isLoadingValues, valuesData, error]);
+
+  const getValuesForLabel = useCallback(
```

Comment:
Could we move this function to the `useCombinedLabels` hook and export it from there?
My main rationale is to try to keep the logic responsible for getting labels and their values in a single place

Reply:
I agree. I've just pushed a commit moving the whole logic for getting labels inside this hook